### PR TITLE
Refactor component to recipe

### PR DESCRIPTION
**Pull request by awaelchli**
Refactors the component structure to a recipe. Based on feedback, more code needs to be exposed so that it is clear how each part can be modified (data, model, trainer, optimization, optimizer, etc.).

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -2,46 +2,88 @@
...
+    def validation_step(self, batch, batch_idx):
+        loss, outputs = self.model(**batch)
+        self.log("val_loss", loss, prog_bar=True)
+
+    def configure_optimizers(self):
+        return AdamW(self.parameters(), lr=0.0001)
+
+
+class MyTextClassification(L.LightningWork):
+    num_classes = 5
```
Comment (justusschock) [inline app.py:53]: nit: still think this belongs to `get_dataset`

**Review by justusschock — APPROVED** [review]

**Comment by edenlightning** [comment]
1. For data- we are still not giving users control over the dataloader, i dont think it's the csv control hat we need here.

I think what we are looking for is more like
```
class MyTextClassification(TextClf):

    def get_dataloader(self):
        train_dataloader = TextClassificationDataLoader(YelpReviewFull(csv_file=os.path.expanduser("~/data/yelpreviewfull/train.csv"))
        val_dataloader = TextClassificationDataLoader(YelpReviewFull(csv_file=os.path.expanduser("~/data/yelpreviewfull/test.csv"))
        return train_dataloader, val_dataloader
	
    def run(self):
        train_dataloader, val_dataloader = self.get_dataloader()
        module, tokenizer = self.get_model(num_labels)
        pl_module = TextClassification(model=module, tokenizer=tokenizer)
        datamodule = TextClassificationData(
            train_dataloader=train_dataloader, val_dataloader=val_dataloader, tokenizer=tokenizer
        )
        trainer = L.Trainer(strategy="deepspeed_stage_3_offload", precision=16)

        trainer.fit(pl_module, datamodule)

```

2. As for finetuning- is there a way to only expose the training_step without the entire LightningModule? in the call you mentioned users might want to change the loss- how would they do that?

File: app.py [hunk 2]
```diff
@@ -2,45 +2,69 @@
...
+        self.tokenizer = tokenizer
+
+    def training_step(self, batch, batch_idx):
+        output = self.model(**batch)
+        self.log("train_loss", output.loss, prog_bar=True, on_epoch=True, on_step=True)
+        return output.loss
+
+    def validation_step(self, batch, batch_idx):
+        output = self.model(**batch)
+        self.log("val_loss", output.loss, prog_bar=True)
```
Comment (Borda) [inline app.py:28]: are we measuring anything more then loss?

File: app.py [hunk 3]
```diff
@@ -2,45 +2,69 @@
...
-    def get_dataset(self):
-        train_dset = YelpReviewFull(csv_file=os.path.expanduser("~/data/yelpreviewfull/train.csv"))
-        val_dset = YelpReviewFull(csv_file=os.path.expanduser("~/data/yelpreviewfull/test.csv"))
-        num_labels = 5
-        return train_dset, val_dset, num_labels
+    def get_dataloaders(self, tokenizer):
+        train_dataloader = TextClassificationDataLoader(
+            dataset=TextDataset(csv_file=os.path.expanduser("~/data/yelpreviewfull/train.csv")),
+            tokenizer=tokenizer,
+        )
```
Comment (Borda) [inline app.py:52]: this formatting looks wild :otter:

File: utilities/callbacks.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import lightning as L
+
+
+def default_callbacks():
```
Comment (Borda) [inline utilities/callbacks.py:4]:
```suggestion
def default_callbacks() -> List[Callbacks]:
```

File: app.py [hunk 5]
```diff
@@ -2,45 +2,69 @@
...
+        val_dataloader = TextClassificationDataLoader(
+            dataset=TextDataset(csv_file=os.path.expanduser("~/data/yelpreviewfull/test.csv")),
+            tokenizer=tokenizer
+        )
+        return train_dataloader, val_dataloader
 
     def get_trainer(self):
```
Comment (Borda) [inline app.py:59]:
```suggestion
    def get_trainer(self) -> Trainer:
```

File: app.py [hunk 6]
```diff
@@ -2,45 +2,69 @@
...
+            model_type, num_labels=self.num_classes, ignore_mismatched_sizes=True
         )
         return model, tokenizer
 
-    def get_dataset(self):
-        train_dset = YelpReviewFull(csv_file=os.path.expanduser("~/data/yelpreviewfull/train.csv"))
-        val_dset = YelpReviewFull(csv_file=os.path.expanduser("~/data/yelpreviewfull/test.csv"))
-        num_labels = 5
-        return train_dset, val_dset, num_labels
+    def get_dataloaders(self, tokenizer):
```
Comment (Borda) [inline app.py:48]:
```suggestion
    def get_dataloaders(self, tokenizer) -> Tuple[Dataloader, Dataloader]:
```

File: utilities/__init__.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+from utilities.callbacks import default_callbacks
```
Comment (Borda) [inline utilities/__init__.py:1]: do we expect this package/repo to be installed alongside other env, as this very generic name may easily collide...

File: utilities/data.py [hunk 8]
```diff
@@ -24,41 +23,24 @@ def __getitem__(self, item):
...
         self,
-        train_dataset: Dataset,
-        val_dataset: Dataset,
+        dataset: Dataset,
         tokenizer: PreTrainedTokenizer,
-        batch_size: int = 4,
         max_token_len: int = 256,
+        batch_size: int = 4,
         num_workers: Optional[int] = None,
+        **kwargs,
```
Comment (Borda) [inline utilities/data.py:34]:
```suggestion
        **kwargs: Any,
```

**Comment by awaelchli** [comment]
@Borda I don't think we will add types to the examples. There are also no strict typing checks on this repo.

**Comment by awaelchli** [comment]
The structure changes were decided by product. The missing typing is intentional and can be added later if we feel like we need it still.
